onnxscript/rewriter/cast_constant_of_shape.py

```diff
@@ -12,58 +13,34 @@
 logger = logging.getLogger(__name__)
 
 
-def cast_constant_of_shape(
-    shape: Sequence[int],
-    t: Any,
-    dtype: int,
-    match_bindings: dict[str, ir.Value | Any] | None = None,
-) -> pattern.OpPattern:
-    constant = op.ConstantOfShape(shape, value=t)
+def cast_constant_of_shape(shape, scalar, dtype):
+    constant = op.ConstantOfShape(shape, value=scalar)
     return op.Cast(constant, to=dtype)
 
 
-def fused_cast_constant_of_shape(
-    shape: Sequence[int], t: Any, dtype: int, match_bindings: dict[str, ir.Value | Any]
-) -> pattern.OpPattern:
-    del dtype  # unused
-    del t  # unused
-    v_dtype = match_bindings["dtype"]
-    v_t = match_bindings["t"]
-    v_dtype = ir.DataType(v_dtype.value).numpy()  # type: ignore[union-attr]
-    casted_val = ir.Tensor(v_t.value.numpy().astype(v_dtype))  # type: ignore[union-attr]
-    return op.ConstantOfShape(shape, value=casted_val)
-
-
-def cast_constant_of_shape_without_value(
-    shape: Sequence[int],
-    dtype: int,
-    match_bindings: dict[str, ir.Value | Any] | None = None,
-) -> pattern.OpPattern:
-    del match_bindings  # Unused
+def fused_cast_constant_of_shape(op, shape: ir.Value, scalar: ir.Attr, dtype: ir.Attr, **_):
+    # Cast scalar (a TensorProto attribute) to the specified dtype
+    scalar_value = scalar.value.numpy().item()
+    cast_value = onnx.helper.make_tensor("value", dtype.value, (), [scalar_value])
```

Comment:
Use ir.Tensor?

Reply:
In fact `scalar.value` should be a Tensor already?

Reply:
I haven't tried your latest changes (today), but ir.Tensor was failing because convert_attributes was expecting a TensorProto (if I remember correctly)

Reply:
I see. It should be fixed now.

Reply:
> In fact `scalar.value` should be a Tensor already?

Yes, but we want to change/cast its type.